// victory-core/lib/victory-legend/victory-legend.js
Object.defineProperty(exports, "__esModule", {
  value: true
});

var _partialRight2 = require("lodash/partialRight");

var _partialRight3 = _interopRequireDefault(_partialRight2);

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _helperMethods = require("./helper-methods");

var _helperMethods2 = _interopRequireDefault(_helperMethods);

var _propTypes3 = require("../victory-util/prop-types");

var _propTypes4 = _interopRequireDefault(_propTypes3);

var _addEvents = require("../victory-util/add-events");

var _addEvents2 = _interopRequireDefault(_addEvents);

var _helpers = require("../victory-util/helpers");

var _helpers2 = _interopRequireDefault(_helpers);

var _victoryLabel = require("../victory-label/victory-label");

var _victoryLabel2 = _interopRequireDefault(_victoryLabel);

var _victoryContainer = require("../victory-container/victory-container");

var _victoryContainer2 = _interopRequireDefault(_victoryContainer);

var _victoryTheme = require("../victory-theme/victory-theme");

var _victoryTheme2 = _interopRequireDefault(_victoryTheme);

var _point = require("../victory-primitives/point");

var _point2 = _interopRequireDefault(_point);

var _border = require("../victory-primitives/border");

var _border2 = _interopRequireDefault(_border);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var fallbackProps = {
  orientation: "vertical",
  width: 450,
  height: 300,
  x: 0,
  y: 0
};

var defaultLegendData = [{ name: "Series 1" }, { name: "Series 2" }];

var VictoryLegend = function (_React$Component) {
  _inherits(VictoryLegend, _React$Component);

  function VictoryLegend() {
    _classCallCheck(this, VictoryLegend);

    return _possibleConstructorReturn(this, (VictoryLegend.__proto__ || Object.getPrototypeOf(VictoryLegend)).apply(this, arguments));
  }

  _createClass(VictoryLegend, [{
    key: "renderChildren",
    value: function renderChildren(props) {
      var _this2 = this;

      var dataComponent = props.dataComponent,
          labelComponent = props.labelComponent,
          title = props.title;

      var dataComponents = this.dataKeys.map(function (_dataKey, index) {
        var dataProps = _this2.getComponentProps(dataComponent, "data", index);
        return _react2.default.cloneElement(dataComponent, dataProps);
      });

      var labelComponents = this.dataKeys.map(function (_dataKey, index) {
        var labelProps = _this2.getComponentProps(labelComponent, "labels", index);
        if (typeof labelProps.text !== "undefined" && labelProps.text !== null) {
          return _react2.default.cloneElement(labelComponent, labelProps);
        }
        return undefined;
      }).filter(Boolean);

      var borderProps = this.getComponentProps(props.borderComponent, "border", 0);
      var borderComponent = _react2.default.cloneElement(props.borderComponent, borderProps);
      if (title) {
        var titleProps = this.getComponentProps(props.title, "title", 0);
        var titleComponent = _react2.default.cloneElement(props.titleComponent, titleProps);
        return [borderComponent].concat(_toConsumableArray(dataComponents), [titleComponent], _toConsumableArray(labelComponents));
      }
      return [borderComponent].concat(_toConsumableArray(dataComponents), _toConsumableArray(labelComponents));
    }
  }, {
    key: "render",
    value: function render() {
      var role = this.constructor.role;

      var props = _helpers2.default.modifyProps(this.props, fallbackProps, role);
      var children = [this.renderChildren(props)];
      return props.standalone ? this.renderContainer(props.containerComponent, children) : _react2.default.cloneElement(props.groupComponent, {}, children);
    }
  }]);

  return VictoryLegend;
}(_react2.default.Component);

VictoryLegend.displayName = "VictoryLegend";
VictoryLegend.role = "legend";
VictoryLegend.propTypes = {
  borderComponent: _propTypes2.default.element,
  borderPadding: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.shape({
    top: _propTypes2.default.number,
    bottom: _propTypes2.default.number,
    left: _propTypes2.default.number,
    right: _propTypes2.default.number
  })]),
  centerTitle: _propTypes2.default.bool,
  colorScale: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.string), _propTypes2.default.oneOf(["grayscale", "qualitative", "heatmap", "warm", "cool", "red", "green", "blue"])]),
  containerComponent: _propTypes2.default.element,
  data: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    name: _propTypes2.default.string.isRequired,
    label: _propTypes2.default.object,
    symbol: _propTypes2.default.object
  })),
  dataComponent: _propTypes2.default.element,
  eventKey: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes4.default.allOfType([_propTypes4.default.integer, _propTypes4.default.nonNegative]), _propTypes2.default.string]),
  events: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    target: _propTypes2.default.oneOf(["data", "labels", "parent"]),
    eventKey: _propTypes2.default.oneOfType([_propTypes2.default.array, _propTypes4.default.allOfType([_propTypes4.default.integer, _propTypes4.default.nonNegative]), _propTypes2.default.string]),
    eventHandlers: _propTypes2.default.object
  })),
  externalEventMutations: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    callback: _propTypes2.default.function,
    childName: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.array]),
    eventKey: _propTypes2.default.oneOfType([_propTypes2.default.array, _propTypes4.default.allOfType([_propTypes4.default.integer, _propTypes4.default.nonNegative]), _propTypes2.default.string]),
    mutation: _propTypes2.default.function,
    target: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.array])
  })),
  groupComponent: _propTypes2.default.element,
  gutter: _propTypes2.default.oneOfType([_propTypes4.default.nonNegative, _propTypes2.default.shape({
    left: _propTypes4.default.nonNegative,
    right: _propTypes4.default.nonNegative
  })]),
  height: _propTypes4.default.nonNegative,
  itemsPerRow: _propTypes4.default.nonNegative,
  labelComponent: _propTypes2.default.element,
  orientation: _propTypes2.default.oneOf(["horizontal", "vertical"]),
  padding: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.shape({
    top: _propTypes2.default.number,
    bottom: _propTypes2.default.number,
    left: _propTypes2.default.number,
    right: _propTypes2.default.number
  })]),
  rowGutter: _propTypes2.default.oneOfType([_propTypes4.default.nonNegative, _propTypes2.default.shape({
    top: _propTypes4.default.nonNegative,
    bottom: _propTypes4.default.nonNegative
  })]),
  sharedEvents: _propTypes2.default.shape({
    events: _propTypes2.default.array,
    getEventState: _propTypes2.default.func
  }),
  standalone: _propTypes2.default.bool,
  style: _propTypes2.default.shape({
    border: _propTypes2.default.object,
    data: _propTypes2.default.object,
    labels: _propTypes2.default.object,
    parent: _propTypes2.default.object,
    title: _propTypes2.default.object
  }),
  symbolSpacer: _propTypes2.default.number,
  theme: _propTypes2.default.object,
  title: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.array]),
  titleComponent: _propTypes2.default.element,
  titleOrientation: _propTypes2.default.oneOf(["top", "bottom", "left", "right"]),
  width: _propTypes4.default.nonNegative,
  x: _propTypes4.default.nonNegative,
  y: _propTypes4.default.nonNegative
};
VictoryLegend.defaultProps = {
  borderComponent: _react2.default.createElement(_border2.default, null),
  data: defaultLegendData,
  containerComponent: _react2.default.createElement(_victoryContainer2.default, null),
  dataComponent: _react2.default.createElement(_point2.default, null),
  groupComponent: _react2.default.createElement("g", null),
  labelComponent: _react2.default.createElement(_victoryLabel2.default, null),
  standalone: true,
  theme: _victoryTheme2.default.grayscale,
  titleComponent: _react2.default.createElement(_victoryLabel2.default, null)
};
VictoryLegend.getBaseProps = (0, _partialRight3.default)(_helperMethods2.default, fallbackProps);
VictoryLegend.expectedComponents = ["borderComponent", "containerComponent", "dataComponent", "groupComponent", "labelComponent", "titleComponent"];
exports.default = (0, _addEvents2.default)(VictoryLegend);
