// victory-core/lib/victory-primitives/common-props.js
Object.defineProperty(exports, "__esModule", {
  value: true
});

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _propTypes3 = require("../victory-util/prop-types");

var _propTypes4 = _interopRequireDefault(_propTypes3);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  active: _propTypes2.default.bool,
  className: _propTypes2.default.string,
  data: _propTypes2.default.array,
  events: _propTypes2.default.object,
  index: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.string]),
  origin: _propTypes2.default.shape({ x: _propTypes2.default.number, y: _propTypes2.default.number }),
  polar: _propTypes2.default.bool,
  role: _propTypes2.default.string,
  scale: _propTypes2.default.oneOfType([_propTypes4.default.scale, _propTypes2.default.shape({ x: _propTypes4.default.scale, y: _propTypes4.default.scale })]),
  shapeRendering: _propTypes2.default.string,
  style: _propTypes2.default.object
};
