// victory-chart/lib/components/victory-scatter/victory-scatter.js
Object.defineProperty(exports, "__esModule", {
  value: true
});

var _partialRight2 = require("lodash/partialRight");

var _partialRight3 = _interopRequireDefault(_partialRight2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _victoryCore = require("victory-core");

var _helperMethods = require("./helper-methods");

var _helperMethods2 = _interopRequireDefault(_helperMethods);

var _commonProps = require("../../helpers/common-props");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var fallbackProps = {
  width: 450,
  height: 300,
  padding: 50,
  size: 3,
  symbol: "circle"
};

var animationWhitelist = ["data", "domain", "height", "maxBubbleSize", "padding", "samples", "size", "style", "width"];

var VictoryScatter = function (_React$Component) {
  _inherits(VictoryScatter, _React$Component);

  function VictoryScatter() {
    _classCallCheck(this, VictoryScatter);

    return _possibleConstructorReturn(this, (VictoryScatter.__proto__ || Object.getPrototypeOf(VictoryScatter)).apply(this, arguments));
  }

  _createClass(VictoryScatter, [{
    key: "shouldAnimate",


    // Overridden in native versions
    value: function shouldAnimate() {
      return !!this.props.animate;
    }
  }, {
    key: "render",
    value: function render() {
      var role = this.constructor.role;

      var props = _victoryCore.Helpers.modifyProps(this.props, fallbackProps, role);
      if (this.shouldAnimate()) {
        return this.animateComponent(props, animationWhitelist);
      }
      var children = this.renderData(props);
      return props.standalone ? this.renderContainer(props.containerComponent, children) : children;
    }
  }]);

  return VictoryScatter;
}(_react2.default.Component);

VictoryScatter.displayName = "VictoryScatter";
VictoryScatter.role = "scatter";
VictoryScatter.defaultTransitions = _victoryCore.DefaultTransitions.discreteTransitions();
VictoryScatter.propTypes = _extends({}, _commonProps.BaseProps, _commonProps.DataProps, {
  bubbleProperty: _propTypes2.default.string,
  maxBubbleSize: _victoryCore.PropTypes.nonNegative,
  minBubbleSize: _victoryCore.PropTypes.nonNegative,
  size: _propTypes2.default.oneOfType([_victoryCore.PropTypes.nonNegative, _propTypes2.default.func]),
  symbol: _propTypes2.default.oneOfType([_propTypes2.default.oneOf(["circle", "diamond", "plus", "square", "star", "triangleDown", "triangleUp"]), _propTypes2.default.func])
});
VictoryScatter.defaultProps = {
  samples: 50,
  scale: "linear",
  standalone: true,
  dataComponent: _react2.default.createElement(_victoryCore.Point, null),
  labelComponent: _react2.default.createElement(_victoryCore.VictoryLabel, null),
  containerComponent: _react2.default.createElement(_victoryCore.VictoryContainer, null),
  groupComponent: _react2.default.createElement("g", null),
  theme: _victoryCore.VictoryTheme.grayscale
};
VictoryScatter.getDomain = _victoryCore.Domain.getDomain.bind(_victoryCore.Domain);
VictoryScatter.getData = _victoryCore.Data.getData.bind(_victoryCore.Data);
VictoryScatter.getBaseProps = (0, _partialRight3.default)(_helperMethods2.default.getBaseProps.bind(_helperMethods2.default), fallbackProps);
VictoryScatter.expectedComponents = ["dataComponent", "labelComponent", "groupComponent", "containerComponent"];
exports.default = (0, _victoryCore.addEvents)(VictoryScatter);
