// victory-chart/lib/helpers/common-props.js
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BaseProps = exports.DataProps = undefined;

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _victoryCore = require("victory-core");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DataProps = exports.DataProps = {
  categories: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.string), _propTypes2.default.shape({
    x: _propTypes2.default.arrayOf(_propTypes2.default.string), y: _propTypes2.default.arrayOf(_propTypes2.default.string)
  })]),
  data: _propTypes2.default.array,
  dataComponent: _propTypes2.default.element,
  labelComponent: _propTypes2.default.element,
  labels: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.array]),
  samples: _victoryCore.PropTypes.nonNegative,
  sortKey: _propTypes2.default.oneOfType([_propTypes2.default.func, _victoryCore.PropTypes.allOfType([_victoryCore.PropTypes.integer, _victoryCore.PropTypes.nonNegative]), _propTypes2.default.string, _propTypes2.default.arrayOf(_propTypes2.default.string)]),
  style: _propTypes2.default.shape({
    parent: _propTypes2.default.object, data: _propTypes2.default.object, labels: _propTypes2.default.object
  }),
  x: _propTypes2.default.oneOfType([_propTypes2.default.func, _victoryCore.PropTypes.allOfType([_victoryCore.PropTypes.integer, _victoryCore.PropTypes.nonNegative]), _propTypes2.default.string, _propTypes2.default.arrayOf(_propTypes2.default.string)]),
  y: _propTypes2.default.oneOfType([_propTypes2.default.func, _victoryCore.PropTypes.allOfType([_victoryCore.PropTypes.integer, _victoryCore.PropTypes.nonNegative]), _propTypes2.default.string, _propTypes2.default.arrayOf(_propTypes2.default.string)]),
  y0: _propTypes2.default.oneOfType([_propTypes2.default.func, _victoryCore.PropTypes.allOfType([_victoryCore.PropTypes.integer, _victoryCore.PropTypes.nonNegative]), _propTypes2.default.string, _propTypes2.default.arrayOf(_propTypes2.default.string)])
};

var BaseProps = exports.BaseProps = {
  animate: _propTypes2.default.oneOfType([_propTypes2.default.bool, _propTypes2.default.object]),
  containerComponent: _propTypes2.default.element,
  domain: _propTypes2.default.oneOfType([_victoryCore.PropTypes.domain, _propTypes2.default.shape({ x: _victoryCore.PropTypes.domain, y: _victoryCore.PropTypes.domain })]),
  domainPadding: _propTypes2.default.oneOfType([_propTypes2.default.shape({
    x: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.arrayOf(_propTypes2.default.number)]),
    y: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.arrayOf(_propTypes2.default.number)])
  }), _propTypes2.default.number, _propTypes2.default.arrayOf(_propTypes2.default.number)]),
  eventKey: _propTypes2.default.oneOfType([_propTypes2.default.func, _victoryCore.PropTypes.allOfType([_victoryCore.PropTypes.integer, _victoryCore.PropTypes.nonNegative]), _propTypes2.default.string]),
  events: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    target: _propTypes2.default.oneOf(["data", "labels", "parent"]),
    eventKey: _propTypes2.default.oneOfType([_propTypes2.default.array, _victoryCore.PropTypes.allOfType([_victoryCore.PropTypes.integer, _victoryCore.PropTypes.nonNegative]), _propTypes2.default.string]),
    eventHandlers: _propTypes2.default.object
  })),
  groupComponent: _propTypes2.default.element,
  height: _victoryCore.PropTypes.nonNegative,
  name: _propTypes2.default.string,
  origin: _propTypes2.default.shape({ x: _propTypes2.default.number, y: _propTypes2.default.number }),
  padding: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.shape({
    top: _propTypes2.default.number, bottom: _propTypes2.default.number,
    left: _propTypes2.default.number, right: _propTypes2.default.number
  })]),
  polar: _propTypes2.default.bool,
  range: _propTypes2.default.oneOfType([_victoryCore.PropTypes.domain, _propTypes2.default.shape({ x: _victoryCore.PropTypes.domain, y: _victoryCore.PropTypes.domain })]),
  scale: _propTypes2.default.oneOfType([_victoryCore.PropTypes.scale, _propTypes2.default.shape({ x: _victoryCore.PropTypes.scale, y: _victoryCore.PropTypes.scale })]),
  sharedEvents: _propTypes2.default.shape({
    events: _propTypes2.default.array,
    getEventState: _propTypes2.default.func
  }),
  standalone: _propTypes2.default.bool,
  theme: _propTypes2.default.object,
  width: _victoryCore.PropTypes.nonNegative
};
