// victory-chart/lib/components/containers/cursor-helpers.js
Object.defineProperty(exports, "__esModule", {
  value: true
});

var _isFunction2 = require("lodash/isFunction");

var _isFunction3 = _interopRequireDefault(_isFunction2);

var _throttle2 = require("lodash/throttle");

var _throttle3 = _interopRequireDefault(_throttle2);

var _victoryCore = require("victory-core");

var _eventHandlers = require("../../helpers/event-handlers");

var _brushHelpers = require("./brush-helpers");

var _brushHelpers2 = _interopRequireDefault(_brushHelpers);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CursorHelpers = {
  onMouseMove: function (evt, targetProps) {
    var onChange = targetProps.onChange,
        dimension = targetProps.dimension,
        domain = targetProps.domain;

    var cursorSVGPosition = _victoryCore.Selection.getSVGEventCoordinates(evt);
    var cursorValue = _victoryCore.Selection.getDataCoordinates(targetProps, targetProps.scale, cursorSVGPosition.x, cursorSVGPosition.y);

    var inBounds = _brushHelpers2.default.withinBounds(cursorValue, {
      x1: domain.x[0],
      x2: domain.x[1],
      y1: domain.y[0],
      y2: domain.y[1]
    });

    if (!inBounds) {
      cursorValue = null;
    }

    if ((0, _isFunction3.default)(onChange)) {
      if (inBounds) {
        var value = dimension ? cursorValue[dimension] : cursorValue;
        onChange(value, targetProps);
      } else if (cursorValue !== targetProps.cursorValue) {
        onChange(targetProps.defaultCursorValue || null, targetProps);
      }
    }

    return [{
      target: "parent",
      eventKey: "parent",
      mutation: function () {
        return { cursorValue: cursorValue };
      }
    }];
  }
};

exports.default = {
  onMouseMove: (0, _throttle3.default)((0, _eventHandlers.attachId)(CursorHelpers.onMouseMove.bind(CursorHelpers)), 32, // eslint-disable-line no-magic-numbers
  { leading: true, trailing: false })
};
